### DISCUSSION TITLE
How do I detect if there are pages to navigate back to

### DISCUSSION THREAD

**Discussion by malonehedges** [vercel/next.js#40383] [Help]
(this is a continuation of a discussion i started in the [wrong place](https://github.com/vercel/next.js/pull/37192#issuecomment-1238670613))

In Next `v12.1.5`, calling `window.history.state.idx` results in `0` if you have freshly loaded the page, and returns a non-zero number if you have been clicking around the website.

This undocumented functionality has been since removed.

I am using this behavior to enable a back button in a website's UI that either
- navigates back 1 page in the browser history (if `idx !== 0`)
- navigates to the home screen (if `idx === 0`)

Is there any other way to detect if you have pages to navigate back to within the website? How should I go about updating this feature when upgrading Next?

**Comment by icyJoseph** [comment]
Hi,

This is not necessarily a Next.js thing, rather, this is a problem with the current History API, and the way we do SPA transitions. 

This video goes in great depth into this issue: https://www.youtube.com/watch?v=cgKUMRPAliw

Most specifically at [10:10](https://youtu.be/cgKUMRPAliw?t=611) the history stack is discussed at great length. In short it is not obvious, nor trivial to know what's next on the stack.

There's a whole new browser API in the works to help us tackle this. There's very little user-land libraries can do to claim a full over arching solution, sort of that a framework could claim as a feature. I guess this is why it was never really documented. 

What we can do is, claim that in our app all navigation ever goes through the router, so that on first load we copy `window.history.state.key` to memory and every time a route change completes, including hash changes I think, we check if we are back at that copied key. Though, just like with `idx` I do believe this is doomed to be unstable and break in the future.

**Comment by malonehedges** [comment]
yea to be clear the solution with checking `=== 0` doesn't work if you refresh the page, it would then take you to the home page with our code currently.

while not idea, this is still a better experience than having the back button send the user back to twitter or wherever they're coming from.

so it sounds like i will have to persist the key somehow and check it is different... hmmm

**Comment by Kasea** [comment]
Having the same issue that @malonehedges is experiencing. Related to the answer from https://github.com/vercel/next.js/issues/2476#issuecomment-850030407 

Not sure how to go about implementing this now... Sure it might've not been perfect, but I would say it accounted for >95% of all use cases using the old idx.
